### PR TITLE
Core ml run phi3.5

### LINKED ISSUE #30: While scripting, usual model wrapper won't work.
For coreML converting, below step is required

1. trace model
2. script model
3. convert to coreML

## trace model

When we trace model it is required to use `torch.jit.trace`. However, LLM utilizes **loop forwarding** inside of forward function. Inside of the generation logic the transformer has `generate` function which supports, beam search and other language generation utilities.

**However, as far as I know model.generate(...) function does not support torch jit tracing.**

Therefore, we have to trace model with `forward()` function, so that we have to implement extra `generate` function to fully generate the sentences and control max output token length.

| I have a question related to `onnx` export, do they also have similar limitations?

related resources: https://discuss.huggingface.co/t/generate-method-for-models-converted-to-torchscript/47728/2

## Script model

After tracing, due to its loop logic to generate sentences, we need to convert traced model into script model to handle `loop` logic. Below is the sample code of loop.

```python

for _ in range(64):
    predictions, _ = model(input_ids)

    token = torch.argmax(predictions[:, -1, :], dim=-1).unsqueeze(1)
    input_ids = torch.cat([input_ids, token], dim=1)

    if token == eos_token:
        break
```
With this code, we cannot control the max_length. I need to figure out how to implement _generate()_ function to handle `max_length`. 

## sample inputs and outputs

**inputs**
```text
<|system|>
You are a helpful assistant.<|end|>
<|user|>
How to explain Internet for a medieval knight?<|end|>
<|assistant|>
```

**outputs**
```text
<|system|> You are a helpful assistant.<|end|><|user|> How to explain Internet for a medieval knight?<|end|><|assistant|> Explaining the Internet to a medieval knight would require a creative approach, as the concept is anachronistic and fundamentally different from the technology and societal structures of the Middle Ages. Here's how you might attempt to convey the idea:

1. **Start with the concept
```

### PR DESCRIPTION
**Pull request by psycoder-sup**
Fixes #30 #31 partially

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/TBD-Labs-AI/latched/blob/repo-setting/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

# Phi3.5 coreML conversion

- fixed) `nan` problem
- comparison between torch traced model's output with coremlModels

coreML models' output:
image

torchscript models' output:
image

The output has slight numerical differences at 1st or 2nd floating point. I assume that this may be occurred due to `int32->int64` conversion problem addressed in #30 .

### CODE REVIEW

File: src/latched/model_wrappers/huggingface.py [hunk 1]
```diff
@@ -3,14 +3,15 @@
...
-
+import torch
+from torch import nn
 from transformers import PreTrainedTokenizer, PreTrainedTokenizerFast
 
 from latched.configs.device import DeviceConfig
 from latched.model_wrappers.base import BaseModelWrapper
 
 HuggingFaceTokenizer: TypeAlias = PreTrainedTokenizer | PreTrainedTokenizerFast
+TracedModule: TypeAlias = torch.jit._trace.TopLevelTracedModule  # type: ignore
```
Comment (sungmanch) [inline src/latched/model_wrappers/huggingface.py:14]: For the future manangement, we need to add `type.py`, however, not needed at now. 
Please reference this: https://github.com/python/mypy/blob/master/mypy/types.py

File: src/latched/model_wrappers/huggingface.py [hunk 2]
```diff
@@ -56,3 +57,31 @@ def _is_llama_model(self, model: nn.Module) -> bool:
         from transformers.models.llama import LlamaModel
 
         return isinstance(model, LlamaModel)
+
+    @property
```
Comment (sungmanch) [inline src/latched/model_wrappers/huggingface.py:61]: Adding this property to the `HuggingFaceModelWrapper` seems not pretty if there were models which aren't using the `eos_token`. How about separate the class like `HuggingFaceLangugaeModelWrapper` and add this property to that class?

**Review by sungmanch — COMMENTED** [review]
Please check the comments

**Review by psycoder-sup — COMMENTED** [review]
@sungmanch hey

File: src/latched/model_wrappers/huggingface.py (hunk 1 above)
Comment (sungmanch) [reply src/latched/model_wrappers/huggingface.py:14]: Yeah, let's consider in the future.

**Review by sungmanch — APPROVED** [review]

**Pull request merged**